File: habit-tracker-frontend/src/app/services/reminder.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Reminder, ReminderCreate, ReminderUpdate } from '../models/reminder.model';

/**
 * Servicio que centraliza las llamadas al backend para los recordatorios.
 * Proporciona metodos para listar, crear y actualizar alarmas de habitos.
 */
@Injectable({
  providedIn: 'root'
})
export class ReminderService {
  private http = inject(HttpClient);
  private apiUrl = 'http://localhost:8080/habit-tracker-backend/api/reminders';

  /**
   * Obtiene la lista de recordatorios de un habito.
   * @param habitId - El id del habito
   * @returns Observable con un array de Reminder
   */
  getRemindersByHabit(habitId: number): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}?habitId=${habitId}`);
  }

  /**
   * Crea un nuevo recordatorio.
   * @param reminder - Los datos del recordatorio a crear
   * @returns Observable con el recordatorio creado
   */
  createReminder(reminder: ReminderCreate): Observable<Reminder> {
    return this.http.post<Reminder>(this.apiUrl, reminder);
  }

  /**
   * Actualiza un recordatorio existente.
   * @param id - El id del recordatorio a actualizar
   * @param reminder - Los datos actualizados del recordatorio
   * @returns Observable con el recordatorio actualizado
   */
  updateReminder(id: number, reminder: ReminderUpdate): Observable<Reminder> {
    return this.http.put<Reminder>(`${this.apiUrl}/${id}`, reminder);
  }
}

File: habit-tracker-backend/src/main/java/com/habittracker/service/ReminderService.java
package com.habittracker.service;

import com.habittracker.dao.ReminderDAO;
import com.habittracker.dto.ReminderDTO;
import com.habittracker.model.Reminder;

import java.time.LocalTime;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio con la logica de negocio para la gestion de recordatorios.
 * Se encarga de crear, listar, actualizar y eliminar recordatorios de habitos.
 */
public class ReminderService {

    private final ReminderDAO reminderDAO;

    public ReminderService() {
        this.reminderDAO = new ReminderDAO();
    }

    /**
     * Crea un nuevo recordatorio a partir de los datos del DTO.
     *
     * @param dto Datos del recordatorio a crear
     * @return ReminderDTO con los datos del recordatorio creado
     */
    public ReminderDTO createReminder(ReminderDTO dto) {
        Reminder recordatorio = new Reminder();
        recordatorio.setHabitId(dto.getHabitId());
        recordatorio.setUserId(dto.getUserId());
        recordatorio.setReminderTime(parsearHora(dto.getReminderTime()));
        recordatorio.setDaysOfWeek(dto.getDaysOfWeek());

        Reminder recordatorioCreado = reminderDAO.save(recordatorio);

        if (recordatorioCreado == null) {
            throw new RuntimeException("Error al crear el recordatorio");
        }

        return new ReminderDTO(recordatorioCreado);
    }

    /**
     * Obtiene todos los recordatorios de un habito.
     *
     * @param habitId ID del habito
     * @return Lista de ReminderDTO
     */
    public List<ReminderDTO> getRemindersByHabit(int habitId) {
        List<Reminder> recordatorios = reminderDAO.findByHabitId(habitId);
        return recordatorios.stream()
                .map(ReminderDTO::new)
                .collect(Collectors.toList());
    }

    /**
     * Actualiza un recordatorio existente.
     *
     * @param dto Datos actualizados del recordatorio
     * @return ReminderDTO con los datos actualizados, o null si no existe
     */
    public ReminderDTO updateReminder(ReminderDTO dto) {
        Reminder recordatorio = new Reminder();
        recordatorio.setId(dto.getId());
        recordatorio.setUserId(dto.getUserId());
        recordatorio.setReminderTime(parsearHora(dto.getReminderTime()));
        recordatorio.setDaysOfWeek(dto.getDaysOfWeek());
        recordatorio.setIsActive(dto.getIsActive());

        boolean actualizado = reminderDAO.update(recordatorio);

        if (!actualizado) {
            return null;
        }

        return new ReminderDTO(recordatorio);
    }

    /**
     * Elimina un recordatorio de la base de datos.
     *
     * @param reminderId ID del recordatorio a eliminar
     * @return true si se elimino correctamente
     */
    public boolean deleteReminder(int reminderId) {
        return reminderDAO.delete(reminderId);
    }

    /**
     * Convierte un String con formato HH:mm a LocalTime.
     *
     * @param hora Hora en formato HH:mm
     * @return LocalTime correspondiente
     */
    private LocalTime parsearHora(String hora) {
        try {
            return LocalTime.parse(hora);
        } catch (DateTimeParseException e) {
            throw new IllegalArgumentException("Formato de hora invalido. Use HH:mm");
        }
    }
}
